### PR TITLE
DP-26913: Fixes running BackstopJS locally in DDEV, updates documentation, and adds a CircleCI job to capture screenshots nightly

### PR DESCRIPTION
**Pull request by justafish**
**Description:**
- Adds a DDEV command to allow BackstopJS to be run locally
- Increases shm_size to 2gb to avoid the browser crashing inside a docker container. See https://bugs.chromium.org/p/chromium/issues/detail?id=519952 and https://github.com/SeleniumHQ/docker-selenium#--shm-size2g
- Switches CircleCI to using backstop's provided image
- Updated the documentation and adds instructions for running the backstop job using CircleCI's local CLI
- Adds a CircleCI job which collects Backstop reference images nightly and stores them as an artifact
- Adds an additional backstop test during the deploy_cd CirlceCI job which uses reference images from the above job

**Jira:** (Skip unless you are MA staff)
DP-26913

**To Test:**
- [ ] Follow the instructions in the README for running things locally
- [ ] Verify examples running in CircleCI for the reference and [test](https://app.circleci.com/pipelines/github/massgov/openmass/23550/workflows/3a5f21fc-e5a3-4e86-90e4-deea48b614bc/jobs/100453) jobs from earlier in this PR
- [ ] Run `ddev drush ma:ci:backstop-snapshot prod --ci-branch=dp-26913-run-backstop-js-locally` successfully (this might take a few tries as production still has timeout issues trying to gather reference images)
- [ ] Run `ddev drush ma:ci:backstop-compare --reference=prod --target=test --ci-branch=dp-26913-run-backstop-js-locally` to test against the newly generated reference images

---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -442,45 +442,137 @@ jobs:
...
+              exit 1
+            fi
+            STOPPED_AT=$(curl --request GET \
+              --fail \
+              --url https://circleci.com/api/v2/project/gh/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/job/${!BACKSTOP_REF} \
+              --header "Circle-Token: $CIRCLE_PERSONAL_TOKEN" | jq -r '.stopped_at')
+            LAST_SUCCESS=$(date -d "$STOPPED_AT" +%s)
+            STALE_TIME=$(date -d "-25 hours" +%s)
+            if [ $LAST_SUCCESS -lte $STALE_TIME ]; then
+              echo "Reference images are stale, please re-run the backstop_reference job"
```
Comment (weitzman) [inline .circleci/config.yml]: Its hard to rerun just one job in CircleCI v2. Lets suggest running nightly_backstop_snapshot workflow instead, and ideally lets create a drush command that will trigger it. I dont think we have such a command today. I can do that if you dont have much Drush command authoring experience.

File: .circleci/config.yml [hunk 2]
```diff
@@ -442,45 +442,137 @@ jobs:
...
+              exit 1
+            fi
+      - run:
+          name: Restore Artifacts
+          command: |
+            BACKSTOP_REF="BACKSTOP_REF_<< parameters.reference >>"
+            URL=$(curl --request GET \
+              --fail \
+              --url https://circleci.com/api/v2/project/gh/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/${!BACKSTOP_REF}/artifacts \
+              --header "Circle-Token: $CIRCLE_PERSONAL_TOKEN" | jq -r '.items | .[] | select(.path=="backstop-prod-latest") | .url')
```
Comment (weitzman) [inline .circleci/config.yml]: Should we be hardcoding env name in `backstop-prod`?

File: .circleci/config.yml [hunk 3]
```diff
@@ -442,45 +442,137 @@ jobs:
...
+            fi
+            STOPPED_AT=$(curl --request GET \
+              --fail \
+              --url https://circleci.com/api/v2/project/gh/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/job/${!BACKSTOP_REF} \
+              --header "Circle-Token: $CIRCLE_PERSONAL_TOKEN" | jq -r '.stopped_at')
+            LAST_SUCCESS=$(date -d "$STOPPED_AT" +%s)
+            STALE_TIME=$(date -d "-25 hours" +%s)
+            if [ $LAST_SUCCESS -lte $STALE_TIME ]; then
+              echo "Reference images are stale, please re-run the backstop_reference job"
+              exit 1
```
Comment (weitzman) [inline .circleci/config.yml:562]: I think this should be just a warning and should not fail the job. Its OK if we use references that are a few days out of date.

File: .ddev/commands/backstop/backstop [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+## Description: Run backstop visual regression tests
+## Usage: backstop [flags] [args]
+## Example: ddev backstop test --target=local --list=all
+
+backstop --config=/src/backstop.js "$@"
```
Comment (weitzman) [inline .ddev/commands/backstop/backstop:7]: When I run this, I get error below. I dont think this is related to my M1. Also, FYI M1 should work with latest Docker Desktop pref as per https://github.com/garris/BackstopJS/issues/1300#issuecomment-1424888520. Thats what I'm trying to test.

```
❯ ddev backstop --help
/bin/sh: 1: /mnt/ddev_config/commands/backstop/backstop: not found
Failed to run backstop --help: exit status 127
```

**Review by weitzman — COMMENTED** [review]
Looks great.

File: .circleci/config.yml (hunk 2 above)
Comment (justafish) [reply .circleci/config.yml]: no! :smile:

File: .circleci/config.yml (hunk 3 above)
Comment (justafish) [reply .circleci/config.yml:562]: I've increased this to a week :+1:

File: .ddev/commands/backstop/backstop (hunk 4 above)
Comment (justafish) [reply .ddev/commands/backstop/backstop:7]: I think your `.ddev/docker-compose.backstop.yaml` file hasn't been replaced with the new one in `.disabled-services/docker-compose.backstop.yaml`

File: drush/Commands/DeployCommands.php [hunk 5]
```diff
@@ -95,6 +95,95 @@ public function backstop(string $target, string $reference, array $options = ['c
...
+   * @option ci-branch The branch that CircleCI should check out at start, default value is "develop"
+   * @usage drush ma:ci:backstop-snapshot --target=prod
+   *   Create a backstop snapshot from production.
+   * @aliases ma-ci-backstop-snapshot
+   * @validate-circleci-token
+   *
+   * @throws \Exception
+   * @throws \GuzzleHttp\Exception\GuzzleException
+   */
+  public function ci_backstop_snapshot(array $options = ['target' => 'prod', 'ci-branch' => 'develop', 'list' => 'all', 'viewport' => 'all']): void {
```
Comment (justafish) [inline drush/Commands/DeployCommands.php:120]: @weitzman I don't know if you'd prefer this in a separate file or not, but I stuck it in here as it's where the existing backstop command is

File: backstop/README.md [hunk 6]
```diff
@@ -1,46 +1,96 @@
...
+- If testing an Acquia environment, make sure the `LOWER_ENVIR_AUTH_USER` and
+`LOWER_ENVIR_AUTH_PASS` environment variables are set up in your `.env` file.
+- Enable the Backstop docker image for local running.
+  `ddev service enable backstop`
+- Before doing any testing, you will need to capture the "reference" screenshots,
+  or the screenshots you want to use as the baseline for comparison. These
+  reference screenshots will usually come from the production environment. Take
+  screen captures of production pages with
+  `ddev backstop reference --target=prod --list=all`
+  Reference screenshots can also be captued from a Tugboat environment with
```
Comment (weitzman) [inline backstop/README.md]:
```suggestion
  Reference screenshots can also be captured from a Tugboat environment with
```

File: backstop/README.md [hunk 7]
```diff
@@ -1,46 +1,96 @@
...
-ahoy backstop test -- --target=local
-# Or use a Tugboat URL as target
-ahoy backstop test -- --target=tugboat --tugboat=https://pr1109-3njbhyjchfr06gmuwiojlcdcxrimpaox.tugboat.qa/
-# Open the report for viewing:
-open backstop/report/index.html
-```
+## Usage
+
+> 🛑 If you're using an M1 Mac, the ddev commands will not work. You can try the
+> local circleci runner as described below.
```
Comment (weitzman) [inline backstop/README.md]:
```suggestion
> 🛑 If you're using an M1 Mac, the ddev commands may not work. You can try the
> local circleci runner as described below, or [turn on Rosetta2 virtualization in Docker Desktop](https://levelup.gitconnected.com/docker-on-apple-silicon-mac-how-to-run-x86-containers-with-rosetta-2-4a679913a0d5)
```

**Review by weitzman — COMMENTED** [review]
PR looks good to me. I'm OK with merging this now, or waiting for the cache bust bypass to be added.

**Pull request merged**